### PR TITLE
Restart/stop the services on upgrade and removal

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
The `flight-web-suite` packages are now `restarted` on `upgrade` and `stopped` on `removal`.

This is done using the `package-scripts`:

* `postinst`: Checks if the service was already running and `restarts` when required,
* `prerm`: Stops the service on full removal, and
* `postrm`: Reloads the `nginx` config.

The `prerm` script is required because `flight-service` can not stop a service it does not know about. This means the services need to be stopped before uninstalling them. This only runs when the package is being removed as opposed to upgraded.

Both the `postinst` and `postrm` scripts need to reload the `nginx` configs. On `upgrade`, `yum` sometimes (always?) runs the `postinst` of the new package before completely removing the older package. This leads to the wrong `nginx` configurations being reloaded. Reloading the `nginx` config in `postrm` fixes this.

---

Broadly the `postinst`/`prerm`/`postrm` scripts are the same for all the `web-suite` packages. A new `update_web_suite_package_scripts.rb` shared software config has been added to manage these files. It is broadly based on `update_puma_scripts.rb`.

The major differences between the difference services are within the `postinst` scripts. It is fairly common to run some form of "configuration" before starting the service. To handle this, a new `postinst-configure` script can be optionally rendered into `postinst`. All custom configuration should be preformed within `postinst-configure` and the build will automatically update `postinst`.

### CODE REVIEW

**Review by benarmston — CHANGES_REQUESTED** [review]
I've made my first pass through this.  I've not yet built or tested the packages, so more comments may be added when I do.

In addition to the comments below: why have the permissions on the `service/types/*/*` scripts changed?  `flight-service` executes those files as `bash /path/to/file` so they don't need to be executable.

File: builders/flight-job-script-api/opt/flight/etc/service/types/job-script-api/start.sh [hunk 1]
```diff
@@ -33,16 +34,13 @@ if [ -f /etc/locale.conf ]; then
...
-log_file="${flight_ROOT}"/var/log/job-script-api/puma.log
-mkdir -p $(dirname "${log_file}")
-
-pidfile=$(mktemp /tmp/flight-deletable.XXXXXXXX.pid)
+# Create the temporary PID file
+pidfile=$(mktemp /tmp/flight-job-script-api-deletable.XXXXXXXX.pid)
 rm "${pidfile}"
 
-tool_bg bash "${flight_ROOT}"/opt/job-script-api/bin/start "$pidfile"
+tool_bg /opt/flight/opt/job-script-api/bin/start "$pidfile"
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply builders/flight-job-script-api/opt/flight/etc/service/types/job-script-api/start.sh:41]: The `puma` service files now managed by a shared `update_puma_scripts`. So the short answer is: _to be consistent with the other `puma` scripts_.

As to why `/opt/flight` is hard coded in general, this is a bit more complicated. 

---

Firstly `flight_ROOT` does not always work everywhere as it won't be defined. In particular: `package-scripts` (e.g. `postinst`), crontabs, and `flight-service` env files.

They would either have to source a file within `/o/f/b` negating the point of `flight_ROOT`. Or use `XDG` which is another can of worms.

This doesn't *technically* apply to the `puma` scripts. However we are currently relying on memory for when it does and does not work.

---

Secondly, using `flight_ROOT` does not add portability in this situation. These scripts are rendered dynamically `update_puma_scripts`. Changing the `/opt/flight` prefix (within the script) would be a reasonable trivial tasks.

There are a few details that would need addressing, nothing consequential:
* The mechanism doesn't currently exist,
* The relative paths within `builders/<app>/opt/flight` would be wrong and
* The files are checked into the repo.

We haven't had a need to move the builds (_yet_), so I've kicked that can down the road. Updating the `puma` scripts would be the least of the concerns.

---

By the time you are running the `start.sh` script, `flight_ROOT` has already been set. This is because it is buried within the `/opt/flight` tree:

`/opt/flight/etc/service/types/job-script-api/start.sh`

It's also extremely difficult to run the `start.sh` script manually due to the special `flight-service` methods (e.g. `tool_bg`). This means there is no use case for the following:

```
# Somehow get tool_bg define
flight_ROOT=/some/where/else /opt/flight/etc/service/types/job-script-api/start.sh
```

I do however expect the following script to start the version the `embedded` version of the code:

```
# This should start the version within /o/f/o/<app>
/opt/flight/opt/<puma-app>/bin/start
```

---

There is also precedence for hard coding the paths within the rpms:

```
[vagrant@chead1 ~]$ cat /opt/flight/opt/service/vendor/ruby/2.7.0/gems/ffi-1.11.1/lib/ffi_c.so  | grep '/opt/flight'
Binary file (standard input) matches
[vagrant@chead1 ~]$ cat /opt/flight/opt/runway/embedded/lib/libruby.so | grep "/opt/flight"
Binary file (standard input) matches
```

This a new `/opt/flight` couldn't be created with a simple `cp`. Instead a new build would be required.

---

TL;DR;

1. We are relying on `memory` for when `flight_ROOT` does and does not work,
2. Hard coding improves consistency and readability to the scripts,
3. There is precedence for hard coding `/opt/flight` (within the builder), and
4. Updating the `puma` script at build time is fairly trivial.

PS: This only applies to files checked into the builder repo. External packages definitely should use `flight_ROOT`.

File: builders/flight-desktop-restapi/config/projects/flight-desktop-restapi.rb [hunk 2]
```diff
@@ -61,8 +62,15 @@
 
 if ohai['platform_family'] == 'rhel'
   runtime_dependency 'flight-desktop >= 1.4.0'
+
+  # TODO: Either remove the lower ~rc* bound on release OR make a flight-service-system-1.1
+  #       The postinst script requires `flight service configure --force`
+  runtime_dependency 'flight-service >= 1.3.0~'
 elsif ohai['platform_family'] == 'debian'
   runtime_dependency 'flight-action-api (>= 1.4.0)'
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply builders/flight-desktop-restapi/config/projects/flight-desktop-restapi.rb]: On review, `flight-desktop-restapi` does not have any configuration options, so `flight service configure --force` is a moot point.

Removed the dep: baa9e8c8c976cce34cb90a3c8d85bffca40674e9

[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply builders/flight-desktop-restapi/config/projects/flight-desktop-restapi.rb]: Oh you are talking about the `flight-action-api`! Didn't even see it

This a pre-existing issue since: 68cbf98a

Fixed: 2edb4cd2ac825d6d67de80a06a1e62037dbd93b5

File: builders/flight-www/package-scripts/flight-www/postinst [hunk 3]
```diff
@@ -71,4 +71,7 @@ EOF
     /opt/flight/bin/flight www cert-gen --config-only ${saved_config}
 fi
 
+# Attempt to reload the nginx configuration
+/opt/flight/bin/flight service reload www >/dev/null 2>&1
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply builders/flight-www/package-scripts/flight-www/postinst:78]: https://github.com/openflighthpc/openflight-omnibus-builder/pull/134/commits/398ba653214bbf54a9472bf6e8613f7025b4bb0a

[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply builders/flight-www/package-scripts/flight-www/postinst:78]: https://github.com/openflighthpc/openflight-omnibus-builder/pull/134/commits/258d9bad4d79608c6214fcd7bbdc6215c724dc97

File: builders/flight-desktop-restapi/config/projects/flight-desktop-restapi.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (benarmston) [reply builders/flight-desktop-restapi/config/projects/flight-desktop-restapi.rb]: :+1:

File: builders/flight-www/package-scripts/flight-www/postinst (hunk 3 above)
[the remark this answers is not in the document]
Comment (benarmston) [reply builders/flight-www/package-scripts/flight-www/postinst:78]: :+1:

File: builders/flight-job-script-api/opt/flight/etc/service/types/job-script-api/start.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (benarmston) [reply builders/flight-job-script-api/opt/flight/etc/service/types/job-script-api/start.sh:41]: You're right that we can't always rely on `flight_ROOT` being set.  But rather than introducing additional hard coding of `/opt/flight` we should use `${flight_ROOT:-/opt/flight}`.

We may never need to relocate where we install our packages, but we shouldn't be making it harder for us to do so if we ever need it.

**Comment by WilliamMcCumstie** [comment]
Rebased to remove rogue `.env` file

File: builders/flight-job-script-api/opt/flight/etc/service/types/job-script-api/start.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply builders/flight-job-script-api/opt/flight/etc/service/types/job-script-api/start.sh]: I've reworked the `puma` scripts to use `flight_ROOT` whether possible. I have avoided using `${flight_ROOT:-/opt/flight}` in fair of exiting with an error. I've done this for two reasons:

1. There is already precedence in favour of an error, and
2. `flight_ROOT` should already be set by `flight-service`, if not, its an error condition.

---

This package scripts are a little bit different. I am fairly certain that `rpm` and `deb` run the scripts in a "clean" environment. This means `flight_ROOT` "should" always be unset.

Using the conditional assignment (`${flight_ROOT:-/opt/flight}`) would apply otherwise, and lead to confusion. Instead I have hard set `flight_ROOT=/opt/flight` at the top of the scripts.

This has two advantages:

* It makes it clear what `flight_ROOT` is within the package scripts, and
* It will be fairly easy to change at a future point if need be

File: builders/flight-console-api/package-scripts/flight-console-api/postinst [hunk 4]
```diff
@@ -56,5 +63,15 @@ if [ ! -f "$pub_key" ]; then
   ssh-keygen -y -f "$priv_key" > "$pub_key"
 fi
 
-/opt/flight/bin/flight service reload www 1>/dev/null 2>&1
+# Added for best practice
+${flight_ROOT}/bin/flight service configure --force >/dev/null 2>&1
```
Comment (benarmston) [inline builders/flight-console-api/package-scripts/flight-console-api/postinst:67]: Looks like this is missing a service name.

Also, my initial testing shows that `--config '{}'` is needed.

```
[vagrant@chead1 (dev) flight-job]$ ${flight_ROOT}/bin/flight service configure www --force  | cat
flight service: Cannot continue within a non-interactive terminal!
Please specify the configs with the following flag: --config JSON
[vagrant@chead1 (dev) flight-job]$ ${flight_ROOT}/bin/flight service configure www --force --config '{}'  | cat
Changes applied.
```

File: builders/flight-console-api/package-scripts/flight-console-api/postinst-configure [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+fi
+
+# Generate the public key if required
+if [ ! -f "$pub_key" ]; then
+  mkdir -p $(dirname "$pub_key")
+  ssh-keygen -y -f "$priv_key" > "$pub_key"
+fi
+
+# Added for best practice
+${flight_ROOT}/bin/flight service configure --force >/dev/null 2>&1
```
Comment (benarmston) [inline builders/flight-console-api/package-scripts/flight-console-api/postinst-configure:32]: Above comments apply here.

File: builders/flight-desktop-restapi/package-scripts/flight-desktop-restapi/postinst [hunk 6]
```diff
@@ -25,6 +25,22 @@
...
-/opt/flight/bin/flight service reload www 1>/dev/null 2>&1
+
+# NOTE: This file is managed by 'update_web_suite_package_scripts' software
+#       definition. Any changes will be lost on the next build.
+
+# Ensure flight_ROOT is set correctly
+flight_ROOT=/opt/flight
+
+# Run the configuration
+${flight_ROOT}/bin/flight service configure --force desktop-restapi >/dev/null 2>&1
```
Comment (benarmston) [inline builders/flight-desktop-restapi/package-scripts/flight-desktop-restapi/postinst:36]: I think `--config '{}'` is needed here.

File: builders/flight-file-manager-api/package-scripts/flight-file-manager-api/postinst [hunk 7]
```diff
@@ -25,15 +25,32 @@
...
+old=${flight_ROOT}/usr/share/file-manager-api
+new=${flight_ROOT}/var/lib/file-manager-api
 if ls "$old" >/dev/null 2>&1; then
   mkdir -p "$new" >/dev/null
   ls "$old" | xargs -I {} mv "$old/{}" "$new"
   rmdir "$old"
 fi
 
+# Added for best practive
+${flight_ROOT}/bin/flight service configure --force file-manager-api >/dev/null 2>&1
```
Comment (benarmston) [inline builders/flight-file-manager-api/package-scripts/flight-file-manager-api/postinst:46]: I think `--config '{}'` is needed here.

File: builders/flight-file-manager-api/package-scripts/flight-file-manager-api/postinst-configure [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+old=${flight_ROOT}/usr/share/file-manager-api
+new=${flight_ROOT}/var/lib/file-manager-api
+if ls "$old" >/dev/null 2>&1; then
+  mkdir -p "$new" >/dev/null
+  ls "$old" | xargs -I {} mv "$old/{}" "$new"
+  rmdir "$old"
+fi
+
+# Added for best practive
+${flight_ROOT}/bin/flight service configure --force file-manager-api >/dev/null 2>&1
```
Comment (benarmston) [inline builders/flight-file-manager-api/package-scripts/flight-file-manager-api/postinst-configure:11]: I think `--config '{}'` is needed here.

File: builders/flight-job-script-api/package-scripts/flight-job-script-api/postinst [hunk 9]
```diff
@@ -25,6 +25,22 @@
...
-/opt/flight/bin/flight service reload www 1>/dev/null 2>&1
+
+# NOTE: This file is managed by 'update_web_suite_package_scripts' software
+#       definition. Any changes will be lost on the next build.
+
+# Ensure flight_ROOT is set correctly
+flight_ROOT=/opt/flight
+
+# Run the configuration
+${flight_ROOT}/bin/flight service configure --force job-script-api >/dev/null 2>&1
```
Comment (benarmston) [inline builders/flight-job-script-api/package-scripts/flight-job-script-api/postinst:36]: I think `--config '{}'` is needed here.

File: builders/flight-login-api/package-scripts/flight-login-api/postinst [hunk 10]
```diff
@@ -25,12 +25,29 @@
...
+
+# Run the configuration
+secret_file=${flight_ROOT}/etc/shared-secret.conf
 if [ ! -f "${secret_file}" ] ; then
     date +%s.%N | sha256sum | cut -c 1-40 > "${secret_file}"
     chmod 0400 "${secret_file}"
 fi
 
+# Added for best practice
+${flight_ROOT}/bin/flight service configure --force login-api
```
Comment (benarmston) [inline builders/flight-login-api/package-scripts/flight-login-api/postinst:43]: I think `--config '{}'` is needed here.

File: builders/flight-www/package-scripts/flight-www/postinst [hunk 11]
```diff
@@ -71,4 +71,10 @@ EOF
     /opt/flight/bin/flight www cert-gen --config-only ${saved_config}
 fi
 
+# Ensure any custom configurations are applied
+/opt/flight/bin/flight service configure --force www >/dev/null 2>&1
```
Comment (benarmston) [inline builders/flight-www/package-scripts/flight-www/postinst:75]: I think `--config '{}'` is needed here.

File: share/config/software/update_web_suite_package_scripts.rb [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+      prerm:    File.join(root_dir, 'package-scripts', project.name, 'prerm'),
+      postrm:   File.join(root_dir, 'package-scripts', project.name, 'postrm')
+    }
+
+    # Render the postinst script
+    configure_script = File.join(project.package_scripts_path, 'postinst-configure')
+    configure = if File.exists? configure_script
+                  File.read(configure_script).chomp
+                else
+                  "${flight_ROOT}/bin/flight service configure --force #{service} >/dev/null 2>&1"
```
Comment (benarmston) [inline share/config/software/update_web_suite_package_scripts.rb:86]: I think `--config '{}'` is needed here.

**Review by benarmston — CHANGES_REQUESTED** [review]

**Comment by benarmston** [comment]
@WilliamMcCumstie what's the reason for the change to the service script permissions?

**Comment by WilliamMcCumstie** [comment]
The `puma` file permissions is a pre-existing change:

https://github.com/openflighthpc/openflight-omnibus-builder/blob/42bdc7c432b308bdb757ca76585bde62a1125b22/share/config/software/update_puma_scripts.rb#L266

The actual permissions are being changed now due to the scripts being re-rendered

File: share/config/software/update_web_suite_package_scripts.rb (hunk 12 above)
Comment (WilliamMcCumstie) [reply share/config/software/update_web_suite_package_scripts.rb]: https://github.com/openflighthpc/openflight-omnibus-builder/pull/134/commits/13bd21d2b900015f568d025b0411c1030d454b2b

File: builders/flight-www/package-scripts/flight-www/postinst (hunk 11 above)
Comment (WilliamMcCumstie) [reply builders/flight-www/package-scripts/flight-www/postinst:75]: https://github.com/openflighthpc/openflight-omnibus-builder/pull/134/commits/13bd21d2b900015f568d025b0411c1030d454b2b

File: builders/flight-login-api/package-scripts/flight-login-api/postinst (hunk 10 above)
Comment (WilliamMcCumstie) [reply builders/flight-login-api/package-scripts/flight-login-api/postinst]: https://github.com/openflighthpc/openflight-omnibus-builder/pull/134/commits/13bd21d2b900015f568d025b0411c1030d454b2b

File: builders/flight-job-script-api/package-scripts/flight-job-script-api/postinst (hunk 9 above)
Comment (WilliamMcCumstie) [reply builders/flight-job-script-api/package-scripts/flight-job-script-api/postinst]: https://github.com/openflighthpc/openflight-omnibus-builder/pull/134/commits/13bd21d2b900015f568d025b0411c1030d454b2b

File: builders/flight-file-manager-api/package-scripts/flight-file-manager-api/postinst-configure (hunk 8 above)
Comment (WilliamMcCumstie) [reply builders/flight-file-manager-api/package-scripts/flight-file-manager-api/postinst-configure]: https://github.com/openflighthpc/openflight-omnibus-builder/pull/134/commits/13bd21d2b900015f568d025b0411c1030d454b2b

File: builders/flight-file-manager-api/package-scripts/flight-file-manager-api/postinst (hunk 7 above)
Comment (WilliamMcCumstie) [reply builders/flight-file-manager-api/package-scripts/flight-file-manager-api/postinst]: https://github.com/openflighthpc/openflight-omnibus-builder/pull/134/commits/13bd21d2b900015f568d025b0411c1030d454b2b

File: builders/flight-desktop-restapi/package-scripts/flight-desktop-restapi/postinst (hunk 6 above)
Comment (WilliamMcCumstie) [reply builders/flight-desktop-restapi/package-scripts/flight-desktop-restapi/postinst]: https://github.com/openflighthpc/openflight-omnibus-builder/pull/134/commits/13bd21d2b900015f568d025b0411c1030d454b2b

File: builders/flight-console-api/package-scripts/flight-console-api/postinst-configure (hunk 5 above)
Comment (WilliamMcCumstie) [reply builders/flight-console-api/package-scripts/flight-console-api/postinst-configure]: https://github.com/openflighthpc/openflight-omnibus-builder/pull/134/commits/13bd21d2b900015f568d025b0411c1030d454b2b

File: builders/flight-console-api/package-scripts/flight-console-api/postinst (hunk 4 above)
Comment (WilliamMcCumstie) [reply builders/flight-console-api/package-scripts/flight-console-api/postinst]: https://github.com/openflighthpc/openflight-omnibus-builder/pull/134/commits/13bd21d2b900015f568d025b0411c1030d454b2b
